### PR TITLE
core: combined epoch seconds and nanos changes in TimeProvider

### PR DESCRIPTION
**Pull request by vinodhabib**
Added changes to get the Timestamp in epoch nanoseconds.

Fixes https://github.com/grpc/grpc-java/issues/5494

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/TimeProvider.java [hunk 1]
```diff
@@ -29,7 +29,10 @@ public interface TimeProvider {
   TimeProvider SYSTEM_TIME_PROVIDER = new TimeProvider() {
     @Override
     public long currentTimeNanos() {
-      return TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis());
+      Instant instant = Instant.now();
+      int nanos = instant.getNano();
+      long epochSeconds = instant.getEpochSecond();
+      return Long.parseLong("" + epochSeconds + nanos);
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/TimeProvider.java:35]: This is not correct. We want the value in nanos. Epoch second of 1 (for example) and Instant.getNano of 1 as per your logic is 11 epoch nanos, whereas it should be 1000000001.

Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:35]: We also need to write a unit test for this.

Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/TimeProvider.java]: Fixed along with adding junits

**Review by kannanjgithub — DISMISSED** [review]

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2017 The gRPC Authors
```
Comment (kannanjgithub) [inline core/src/test/java/io/grpc/internal/TimeProviderTest.java:2]: Change it to 2024.

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    // Get the current time from Instant for comparison
+    Instant instantNow = Instant.now();
+    long expectedTimeNanos = instantNow.getEpochSecond() * 1_000_000_000L + instantNow.getNano();
+
+    // Validate the time returned is close to the expected value within a tolerance.
+    long toleranceOfTwoMilliSecond = 2_000_000L; // 2 millisecond tolerance in nanoseconds
+    assertTrue("The current time in nanoseconds should be close to the expected time.",
+        Math.abs(actualTimeNanos - expectedTimeNanos) < toleranceOfTwoMilliSecond);
+  }
```
Comment (kannanjgithub) [inline core/src/test/java/io/grpc/internal/TimeProviderTest.java:42]: Seeing that other tests like https://github.com/grpc/grpc-java/blob/master/api/src/test/java/io/grpc/CallOptionsTest.java#L150 are allowing for a 10ms tolerance time, lets do the same here to avoid the potential for flake with 2ms.
Also for uniformity sake do the assertion using assertAbout and isWithin like in that test.

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java (hunk 2 above)
Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/TimeProviderTest.java]: Fixed

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java (hunk 3 above)
Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/TimeProviderTest.java:42]: Fixed the tolerance time increase part.
Looks like for the usage of assertAbout is the different use case and I can see only in CallOptionsTest and DeadlineTest used in grpc. 
maybe it's not suitable in our case further checking on this.
also I am unable to import static com.google.common.truth.Truth.assertAbout;

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    // Get the current time from the TimeProvider
+    long actualTimeNanos = TimeProvider.SYSTEM_TIME_PROVIDER.currentTimeNanos();
+
+    // Get the current time from Instant for comparison
+    Instant instantNow = Instant.now();
+    long expectedTimeNanos = instantNow.getEpochSecond() * 1_000_000_000L + instantNow.getNano();
+
+    // Validate the time returned is close to the expected value within a tolerance.
+    long toleranceOfTenMilliSecond = 10_000_000L; // 10 millisecond tolerance in nanoseconds
```
Comment (kannanjgithub) [inline core/src/test/java/io/grpc/internal/TimeProviderTest.java:39]: No need for this variable. Just use the value directly in the comparison below.

**Review by kannanjgithub — DISMISSED** [review]

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java (hunk 4 above)
Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/TimeProviderTest.java]: Fixed

File: core/src/main/java/io/grpc/internal/TimeProvider.java [hunk 5]
```diff
@@ -29,7 +29,10 @@ public interface TimeProvider {
   TimeProvider SYSTEM_TIME_PROVIDER = new TimeProvider() {
     @Override
     public long currentTimeNanos() {
-      return TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis());
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Revert this change. This code is used by all callers. We may only call java.time when the caller is using java.time, as java.time is not always available. We learn that java.time is available by someone calling us with a java.time object.

**Review by ejona86 — CHANGES_REQUESTED** [review]
We shouldn't do this.

File: core/src/main/java/io/grpc/internal/TimeProvider.java (hunk 5 above)
Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Can you elaborate how java.time is needed by java.util.concurrent.TimeUnit? The change was done because grpc needs Java 8 and TimeUnit's toNanos is supported in Java 8.

Comment (ejona86) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Oh, I thought this was for #10245. java.time was added in Android API level 26. https://developer.android.com/reference/java/time/package-summary . We are supporting 21. We run animalsniffer to detect cases like this, but apparently it is broken, because this is the third(?) time recently it hasn't detected a problem.

If we want to use java.time, we'd probably want two separate time providers and detect which to use at runtime.

Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Ok, but this PR uses java.util.concurrent and the fact that Java 8 supports nanoseconds (and Android API level being orthogonal to this ability).

Comment (ejona86) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: What uses `java.util.concurrent`? I see `import java.time.Instant;` above. Shared projects like this (vs something non-Android like xds) use the intersection of {Java 8, Android API 21}, as otherwise it won't work on a supported platform. If we really want something that is only on one platform, then generally we need reflection and a fallback.

Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Got it, I misread the package being used because the comment shows in GitHub UI on the left panel (old code).

Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: @vinodhabib we need to avoid having imports for Java.time.Instant. We need to use reflection (such as Class.forName("Java.time.Instant")) to check if the class is available, and if so do the new changes, but **also using reflection**., and if not available, continue to do the old way using java.util.concurrent.TimeUnit and the nanoseconds offset logic that was there before. There won't be any occurrence of the Instant class statically compiled.

Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Calling methods via reflection has a performance overhead (the creation of the Instant object as well as calling the toNanos method both have to be done via reflection everytime TimeProvider.currentTimeNanos is called, we cannot create and keep the Instant object at startup time and do just 1 method call getNanos on this object via reflection everytime the method is called), is it worth to do this change to use Java.time's nanos? Can we leave it as it is?
 @vinodhabib please hold off your changes for now.

Comment (panchenko) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Leaving it as is is definitely an option.
If trying to implement this - can be 2 implementations of `TimeProvider` and `SYSTEM_TIME_PROVIDER` initialised once at the moment of class loading to one or another.

File: core/src/test/java/io/grpc/internal/TimeProviderTest.java (hunk 3 above)
Comment (ejona86) [reply core/src/test/java/io/grpc/internal/TimeProviderTest.java:42]: We tend to prefer Truth to JUnit Assert, as it is generally easier to use with better error messages. gRPC is older than Truth, so there's still plenty of usages of JUnit's Assert. You'd also use Assert for things like assertThrows(), which doesn't have/need a Truth equivalent.

`assertTrue()` in particular is particularly bad because you don't know the values that failed the assertion. And since this is guaranteed to be flaky, we really will want to know how close the values were when it fails.

```java
assertThat(actualTimeNanos).isWithin(10_000_000L).of(actualTimeNanos);
```

Note that 10 ms is very precise for a time-sensitive test. We generally need to run the code twice, the first for class loading, and even then have seen seconds of delay on some CI workers. Some tests we run in a loop of a few times and choose the best result. These tests are run concurrently with other tests and class loading in particular requires I/O.

Since this is wall time, also realize the resolution could be particularly low. [Back in the day](https://web.archive.org/web/20150924165427/http://blogs.oracle.com/dholmes/entry/inside_the_hotspot_vm_clocks), it was widely known that Windows had 10-15 ms resolution. Seems since Windows 8 or 10/Server 2016 more precision is available.

Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/TimeProviderTest.java:41]: Fixed

File: core/src/main/java/io/grpc/internal/TimeProvider.java (hunk 5 above)
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: yet to work on this change.

Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/TimeProvider.java:32]: Fixed

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  public InstantTimeProvider(Class<?> instantClass) {
+    this.instantClass = instantClass;
+  }
+
+  @Override
+  public long currentTimeNanos() {
+    try {
+      Object instant = instantClass.getMethod("now").invoke(null);
+      int nanos = (int) instantClass.getMethod("getNano").invoke(instant);
+      long epochSeconds = (long) instantClass.getMethod("getEpochSecond").invoke(instant);
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/InstantTimeProvider.java:36]: As Eric said https://github.com/grpc/grpc-java/pull/11604/files/58073d47f0127724e8fb7a684b7492276c707598#r1804991178 Getting the method references by reflection has a performance cost so we should store the method references in the constructor and in the _currentTimeNanos_ method only do the invoke operation.

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      return epochSeconds * 1_000_000_000L + nanos;
```
[the remark this answers is not in the document]
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java:37]: I don't think there's a need to check MIN_VALUE, because this is current time; that would mean the Java API itself has overflowed unix time with a `long`, which will be billions of years in the future.

[the remark this answers is not in the document]
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java]: Fixed

File: core/src/main/java/io/grpc/internal/TimeProvider.java [hunk 8]
```diff
@@ -29,7 +29,10 @@ public interface TimeProvider {
...
+      Instant instant = Instant.now();
+      int nanos = instant.getNano();
+      long epochSeconds = instant.getEpochSecond();
+      return epochSeconds * 1_000_000_000L + nanos;
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/TimeProvider.java]: Fixed

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java (hunk 6 above)
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java]: Fixed

File: core/src/main/java/io/grpc/internal/TimeProviderResolverFactory.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+final class TimeProviderResolverFactory {
+
+  private static final Class<?> INSTANT_CLASS = initInstant();
+
+  private static Class<?> initInstant() {
+    try {
+      return Class.forName("java.time.Instant");
+    } catch (Exception ex) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/TimeProviderResolverFactory.java:30]: This entire PR is over-catching. Catch only the exceptions that you need, so ClassNotFoundException here.

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Override
+  public long currentTimeNanos() {
+    try {
+      Object instant = now.invoke(null);
+      int nanos = (int) getNano.invoke(instant);
+      long epochSeconds = (long) getEpochSecond.invoke(instant);
+      return TimeUnit.SECONDS.toNanos(epochSeconds) + nanos;
+    } catch (Exception ex) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/InstantTimeProvider.java:47]: Don't catch all of Exception. There's two checked exceptions that need to be handled. `IllegalAccessException` should not happen, and `InvocationTargetException` should only contain a RuntimeException. So repackaging them as a RuntimeException is fine, but it isn't clear when you catch all exceptions.

File: core/src/main/java/io/grpc/internal/TimeProviderResolverFactory.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+package io.grpc.internal;
+
+/**
+ * {@link TimeProviderResolverFactory} resolves Time providers.
+ */
+
+final class TimeProviderResolverFactory {
+
+  private static final Class<?> INSTANT_CLASS = initInstant();
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/TimeProviderResolverFactory.java:25]: Fixed

File: core/src/main/java/io/grpc/internal/TimeProviderResolverFactory.java (hunk 9 above)
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/TimeProviderResolverFactory.java:30]: Fixed

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  private Method getNano;
+  private Method getEpochSecond;
+
+  public InstantTimeProvider(Class<?> instantClass) {
+    try {
+      this.now = instantClass.getMethod("now");
+      this.getNano = instantClass.getMethod("getNano");
+      this.getEpochSecond = instantClass.getMethod("getEpochSecond");
+    } catch (NoSuchMethodException e) {
+      throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java:36]: Fixed (Option 1)

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java (hunk 10 above)
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java:47]: Fixed

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      // (i,e 10 millisecond tolerance in nanoseconds).
+      assertThat(actualTimeNanos).isWithin(10_000_000L).of(expectedTimeNanos);
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // Test to verify exception is thrown when "now" method is missing
+  @Test(expected = RuntimeException.class)
+  public void testMissingNowMethodThrowsException() {
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java:73]: Removed this UT as it is no more relevant.

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+      // Get the current time from Instant for comparison
+      Instant instantNow = Instant.now();
+      long expectedTimeNanos = TimeUnit.SECONDS.toNanos(instantNow.getEpochSecond())
+          + instantNow.getNano();
+
+      // Validate the time returned is close to the expected value within a tolerance
+      // (i,e 10 millisecond tolerance in nanoseconds).
+      assertThat(actualTimeNanos).isWithin(10_000_000L).of(expectedTimeNanos);
+    } catch (ClassNotFoundException | NoSuchMethodException e) {
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java:66]: No need to catch, just add `throws Exception` to the test method. Any thrown exceptions will cause the test to fail.

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Unit tests for {@link InstantTimeProvider}.
+ */
+@RunWith(JUnit4.class)
+public class InstantTimeProviderTest {
+  @Test
+  public void testConcurrentCurrentTimeNanos() {
+
+    ConcurrentTimeProvider concurrentTimeProvider = new ConcurrentTimeProvider();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java:35]: A bit surprising to be testing ConcurrentTimeProvider in InstantTimeProviderTest.

Although I'd be fine with both implementations being part of TimeProviderResolverFactory and this being TimeProviderResolverFactoryTest. (Or obviously, move it out to its own test class.)

**Review by ejona86 — APPROVED** [review]
Tell us when you're ready for it to be merged.

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java (hunk 14 above)
Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java]: Fixed

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    this.getEpochSecond = instantClass.getMethod("getEpochSecond");
+  }
+
+  @Override
+  public long currentTimeNanos() {
+    try {
+      Object instant = now.invoke(null);
+      int nanos = (int) getNano.invoke(instant);
+      long epochSeconds = (long) getEpochSecond.invoke(instant);
+      return TimeUnit.SECONDS.toNanos(epochSeconds) + nanos;
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/InstantTimeProvider.java:43]: If the return value of TimeUnit.SECONDS.toNanos(epochSeconds) is Long.MAX_VALUE then just return it without adding nanos. No need to check for MIN_VALUE as Eric mentioned.

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    long actualTimeNanos = instantTimeProvider.currentTimeNanos();
+
+    // Get the current time from Instant for comparison
+    Instant instantNow = Instant.now();
+    long expectedTimeNanos = TimeUnit.SECONDS.toNanos(instantNow.getEpochSecond())
+          + instantNow.getNano();
+
+    // Validate the time returned is close to the expected value within a tolerance
+    // (i,e 10 millisecond tolerance in nanoseconds).
+    assertThat(actualTimeNanos).isWithin(10_000_000L).of(expectedTimeNanos);
```
Comment (kannanjgithub) [inline core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java:48]: I hope the concern mentioned in [this comment ](https://github.com/grpc/grpc-java/pull/11604#discussion_r1805055709) that 10ms is very precise and we may need to run the test twice to first load the class doesn't apply anymore now that we are only doing the method invocation in InstantTimeProvider.currentTimeNanos and not the class loading or getting the methods by reflection. @ejona86

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java]: I think option 2 is better because the exception can never occur. With option 1 will be less easier to understand in the future why the fallback code came into existnce.

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java (hunk 16 above)
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java:43]: Even if we do a check for long.MAX_VALUE the probability of meeting "if condition" being TRUE is exactly once and probability of invoking this method in that point in time is negligible.

Referring to Eric's below comment, he is also in the agreement to live with it.

image

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java (hunk 17 above)
Comment (vinodhabib) [reply core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java:48]: @kannanjgithub hope this is the question for Eric and there is no action item from our end, please confirm?

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java (hunk 16 above)
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java:43]: It doesn't change the discussion, but I do want to make clear: The probability _isn't_ exactly once, because TimeUnit.toNanos saturates at MAX_VALUE. It is, "every result of toNanos after some point in 2262 will be MAX_VALUE."

And not that it matters, but I'll take this opportunity to inform y'all that [LongMath.saturateAdd()](https://javadoc.io/doc/com.google.guava/guava/latest/com/google/common/math/LongMath.html#saturatedAdd(long,long)) exists and is convenient when you're doing lots of this sort of thing.

Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java]: Fixed

File: core/src/main/java/io/grpc/internal/InstantTimeProvider.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (vinodhabib) [reply core/src/main/java/io/grpc/internal/InstantTimeProvider.java]: Fixed

**Review by kannanjgithub — APPROVED** [review]

File: core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java (hunk 17 above)
Comment (ejona86) [reply core/src/test/java/io/grpc/internal/InstantTimeProviderTest.java:48]: Class loading still happens. I wasn't referring to reflection when I was talking about class loading. I was talking about using and calling classes, whether via reflection on directly in code.

If this test isn't flaky, it'd probably be because InstantTimeProvider is using the exact same classes that the test itself uses, so there is no loading between when currentTimeNanos() computes the time (it can load earlier in that call without issue) and the `Instant.now()` in this test method.
